### PR TITLE
PHP 8.1 for CS9

### PR DESCRIPTION
**Pull request by xrow**
Hi,

I made php 8.1 build for CS9. Please review.

### CODE REVIEW

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
As far as I know, PHP container images are UBI-based, and therefore CentOS Stream 9 is not needed. The PHP8.1 image can be pull from registry: `registry.access.redhat.com/ubi9/php-81` . The link is here: https://catalog.redhat.com/software/containers/ubi9/php-81/62e8e662f6d3d47e19779b01?container-tabs=gti

**Comment by xrow** [comment]
Hi,

Are you sure that if I want to install additional packages from remi later on that this will work with registry.access.redhat.com/ubi9/php-81?

Maybe a centos build makes sense. Please let me know.

```text
Package php-sodium-8.1.18-1.el9.remi.x86_64 is already installed.
Package php-pecl-zip-1.20.1-1.module+el9.1.0+15477+1fc469b9.x86_64 is already installed.
Error: 
 Problem 1: cannot install the best candidate for the job
  - nothing provides librabbitmq(x86-64) >= 0.11.0 needed by php-pecl-amqp-1.11.0-1.el9.remi.8.1.x86_64
  - nothing provides librabbitmq.so.4()(64bit) needed by php-pecl-amqp-1.11.0-1.el9.remi.8.1.x86_64
 Problem 2: package php-pecl-imagick-im6-3.7.0-7.el9.remi.8.1.x86_64 requires libMagickCore-6.Q16.so.7()(64bit), but none of the providers can be installed
  - package php-pecl-imagick-im6-3.7.0-7.el9.remi.8.1.x86_64 requires libMagickWand-6.Q16.so.7()(64bit), but none of the providers can be installed
  - cannot install the best candidate for the job
  - nothing provides libcgraph.so.6()(64bit) needed by ImageMagick-libs-6.9.12.82-1.el9.x86_64
  - nothing provides libgvc.so.6()(64bit) needed by ImageMagick-libs-6.9.12.82-1.el9.x86_64
  - nothing provides libraw_r.so.20()(64bit) needed by ImageMagick-libs-6.9.12.82-1.el9.x86_64
  - nothing provides libwmflite-0.2.so.7()(64bit) needed by ImageMagick-libs-6.9.12.82-1.el9.x86_64
  - nothing provides libcdt.so.5()(64bit) needed by ImageMagick-libs-6.9.12.82-1.el9.x86_64
  - nothing provides libcgraph.so.6()(64bit) needed by ImageMagick-libs-1:6.9.12.85-1.el9.remi.x86_64
  - nothing provides libgvc.so.6()(64bit) needed by ImageMagick-libs-1:6.9.12.85-1.el9.remi.x86_64
  - nothing provides libraw_r.so.20()(64bit) needed by ImageMagick-libs-1:6.9.12.85-1.el9.remi.x86_64
  - nothing provides libwmflite-0.2.so.7()(64bit) needed by ImageMagick-libs-1:6.9.12.85-1.el9.remi.x86_64
  - nothing provides libcdt.so.5()(64bit) needed by ImageMagick-libs-1:6.9.12.85-1.el9.remi.x86_64
  - nothing provides libcgraph.so.6()(64bit) needed by ImageMagick-libs-1:6.9.12.86-1.el9.remi.x86_64
  - nothing provides libgvc.so.6()(64bit) needed by ImageMagick-libs-1:6.9.12.86-1.el9.remi.x86_64
  - nothing provides libraw_r.so.20()(64bit) needed by ImageMagick-libs-1:6.9.12.86-1.el9.remi.x86_64
  - nothing provides libwmflite-0.2.so.7()(64bit) needed by ImageMagick-libs-1:6.9.12.86-1.el9.remi.x86_64
  - nothing provides libcdt.so.5()(64bit) needed by ImageMagick-libs-1:6.9.12.86-1.el9.remi.x86_64
```

**Comment by xrow** [comment]
@phracek Please share your thoughts on this. Centos has many/some additional prebuild php extensions that rhel seem to not have.

**Comment by pkubatrh** [comment]
Centos Stream is essentially a future version of RHEL. It might have some packages that will be added in the next RHEL version, but I doubt the package set will be that much different.

@remicollet might be able to help with installing additional extensions on top of the image

**Comment by remicollet** [comment]
BTW in 32477ecfb6bef28416a0dc98b25b1d0d1f0217be
The `yum module -y enable php:$PHP_VERSION` is indeed missing in Dockerfile.c9s while present in Dockerfile.rhel9 
(no need to use the `switch-to` option)

**Comment by remicollet** [comment]
@xrow missing dependencies are from EPEL which is a 3rd party repository, not enabled (AFAIK) in any s2i variant.

**Comment by xrow** [comment]
Yes I know I enabled EPEL... Though ubi9 with EPEL was missing the stuff I was laying out before.

**Comment by xrow** [comment]
Hi All, @remicollet 

I had more time now investigating. How much my working c9s stack on remi-8.1 that fully installs would differ from the ubi9 stack and kudos to remi to make this even possible.

My desire is to have those packages wokring.
 
```
RUN dnf module -y enable php:$PHP_VERSION && \
    INSTALL_PKGS="php-pecl-amqp \
                  php-sodium \
                  php-pecl-imagick-im6 \
                  php-pecl-redis5 \
                  php-pecl-http \
                  php-pecl-msgpack \
                  php-pecl-lzf \
                  php-pecl-igbinary \
                  php-pecl-xdebug3 \
                  php-pecl-zip" && \
    dnf install -y $INSTALL_PKGS && \
    rpm -V $INSTALL_PKGS
```

The results are
* with c9s + remi-8.1 + epel all is installable
* with ubi9 + remi-8.1 + epel all is installable, but fails for php-pecl-amqp (missing librabbitmq-0.11 ) and php-pecl-imagick-* (missing a lot of extra stuff)
* with ubi9 without remi-8.1 most of the packages are not pre build.
* didn`t bother to try anything with pecl install on either platform

My question would be what is the most reasonable approach to roll out such a more demanding stack. To be honest to me it looks like enabling c9s as a base and continue with remi looks like the best option to me as long the other packages are not available for the base os / appstream repos.

Would you still think c9s should be still disabled?

**Comment by xrow** [comment]
Hi @phracek,
Hi @remicollet,

I revisited the issue. I agree there is no eminent need to have cs9 because you can archive most things with ubi. So I I did still keep it disabled per default. Nevertheless I fixed the Dockerfiles for cs9 for php 8.1 and 8.2 in case you want to build it. Please accept this pull now.

**Comment by zmiklank** [comment]
@xrow Thank you for the work on this PR.

It looks good to me; your change fixes the build of c9s images - even though the build is still disabled by default, as you already pointed out. 

Could you please squash all commits in this PR into one? I could do that while merging, but it seems there is one merge commit present and I am not sure how the GitHub UI would handle that.

**Comment by xrow** [comment]
@zmiklank 

You can configure squash for the project and PR in github in general. Nevertheless I will rebase. Please check if you can also configure the project so that it will become easier the next time.

https://docs.github.com/de/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

File: 8.1/.exclude-c9s [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (zmiklank) [inline 8.1/.exclude-c9s:1]: This change in .exclude-c9s is not needed.

**Comment by xrow** [comment]
@zmiklank Sorry I fucked up on this by accidently closeing it. Can you merge https://github.com/sclorg/s2i-php-container/pull/437 instead?

**Pull request closed without merging**